scripts/monitor/visualization/resource_monitoring.py

```diff
@@ -0,0 +1,596 @@
+            shards = self.meta_runtime.loc[:, "runtime_shard"]
+            for i in range(0, tasks.size):
+                if tasks[i] == task_name_or_index and shards[i] == shard_index:
+                    pos = i
+        else:
+            if isinstance(task_name_or_index, int):
+                pos = task_name_or_index  # if only task_name is supplied it's treated as the row index
+            else:
+                raise Exception(f"Need an integer row index but received {task_name_or_index}")
+        input_ds = pd.DataFrame(data=self.meta_runtime.loc[pos, "meta_inputs"])
+        return input_ds[input_ds['type'].astype(str) == "file"].loc[:, "value"].astype(float).sum()
+
+    # ''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+    def cpu_use_violin(self, instance_id: int) -> plt.Figure:
+        """
+        Violin plot for CPU use
+        :param instance_id: id of vm a shard is running on
+        :return: plot
+        """
+        by_id = self.metrics[self.metrics["metrics_instance_id"] == instance_id]\
+            .sort_values("metrics_timestamp").loc[:, "metrics_cpu_used_percent"]
+        data = list()
+        for i in range(len(by_id.iloc[0])):
+            data.append(pd.Series(list(e[i] for e in by_id)))
+        fig = plt.figure()
+        axis = fig.add_axes([0, 0, 1, 1])
+        _ = axis.violinplot(data, showmeans=True, showextrema=True, showmedians=True)
+        name = self.get_shard_from_instance_id(instance_id)
+        axis.set_xlabel("CPU core #")
+        axis.set_ylabel("CPU % used")
+        axis.set_title(f"CPU Usage for {name[0]}, shard {name[1]}")
+        plt.show()
+        # plt.savefig(f"{self.cpu_use_violin.__name__}.pdf", bbox_inches='tight')
+        return fig
+
+    def get_instance_id(self, task: str, shard_idx: int) -> numpy.int64:
+        """
+        Gets instance ID given a shard
+        :param task: task name
+        :param shard_idx: shard index #
+        :return: instance ID
+        """
+        tasks = self.meta_runtime.loc[:, "runtime_task_call_name"]  # find row index
+        shards = self.meta_runtime.loc[:, "runtime_shard"]
+        pos = None
+        for i in range(len(shards)):
+            if tasks[i] == task and shards[i] == shard_idx:
+                pos = i
+        try:
+            return self.meta_runtime.loc[pos, "runtime_instance_id"]
+        except KeyError:
+            print(f"task {task} shard {shard_idx} was not found")
+
+    def cpu_plot(self, task: str, shard_idx: int) -> None:
+        """
+        Combining cpu_use_violin and get_instance_id (is this considered a wrapper function? maybe not)
+        :param task: task name
+        :param shard_idx: shard index #
+        :return: plot
+        """
+        self.cpu_use_violin(self.get_instance_id(task, shard_idx))
+
+    def mem_use_violin(self, instance_id: int) -> plt.Figure:
```

Comment:
Given this is sharing code with `cpu_use_violin`, let's see if the refactoring suggested above makes sense to you.
My guess it that will lead to refactor of this as well.